IPv4/IPv6 dual stack docs (#16010)
kubernetes/website: content/en/docs/concepts/cluster-administration/networking.md

or stand-alone.  In either version, it doesn't require any configuration or extra code
to run, and in both cases, the network provides one IP address per pod - as is standard for Kubernetes.

## IPv4/IPv6 dual stack

{{< feature-state state="alpha" >}}

> If you enable IPv4/IPv6 dual stack networking for you Kubernetes cluster, the cluster will support the simultaneous assignment of both IPv4 and IPv6 addresses.

Reviewer: nit:
```suggestion
 If you enable IPv4/IPv6 dual stack networking for your Kubernetes cluster, the cluster will support the simultaneous assignment of both IPv4 and IPv6 addresses.
```

   * kube-controller-manager:
      * `--feature-gates="IPv6DualStack=true"`
      * `--cluster-cidr=<IPv4 CIDR>,<IPv6 CIDR>` eg. `--cluster-cidr=10.244.0.0/16,fc00::/24`
      * `--service-cluster-ip-range=<IPv4 CIDR>,<IPv6 CIDR>`
   * kubelet:
      * `--feature-gates="IPv6DualStack=true"`
   * kube-proxy:
      * `--proxy-mode=ipvs`
      * `--cluster-cidrs=<IPv4 CIDR>,<IPv6 CIDR>` 
      * `--feature-gates="IPv6DualStack=true"`

{{< note >}}
An IPv6 CIDR specificed via the `--cluster-cidr` flag larger than /24 will fail
{{< /note >}}

### Services

If your cluster has IPv4/IPv6 dual-stack networking enabled, you can create {{< glossary_tooltip text="Services" term_id="service" >}} with either an IPv4 or an IPv6 address. You can choose the address family for the Service's cluster IP by setting a field, `.spec.ipFamily`, on that Service.
You can only set this field when creating a new Service. The default address family for your cluster is the address family of the first service cluster IP range configured via the `--service-cluster-ip-range` flag to the kube-controller- manager.

You can set `.spec.ipFamily` to either:

>    * `IPv4`: api-server will assign an IP from a `service-cluster-ip-range` that is `ipv4`

Reviewer: nit: the docs commonly use “the API server”; that seems like the right fit here.

The following prerequisites are needed in order to utilize IPv4/IPv6 dual stack Kubernetes clusters:

   * Kubernetes 1.16 or later
   * Provider support for dual stack networking (Cloud provider or otherwise must be able to provide Kubernetes nodes with routable IPv4/IPv6 network interfaces)
   * Kubenet network plugin
   * Kube-proxy running in mode IPVS

### Enable IPv4/IPv6 dual stack

To enable IPv4/IPv6 dual stack, enable the `IPv6DualStack` [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) for the relevant components of your cluster, and set dual-stack cluster network assignments:

   * kube-controller-manager:
      * `--feature-gates="IPv6DualStack=true"`
      * `--cluster-cidr=<IPv4 CIDR>,<IPv6 CIDR>` eg. `--cluster-cidr=10.244.0.0/16,fc00::/24`
      * `--service-cluster-ip-range=<IPv4 CIDR>,<IPv6 CIDR>`
   * kubelet:
      * `--feature-gates="IPv6DualStack=true"`
   * kube-proxy:
      * `--proxy-mode=ipvs`
      * `--cluster-cidrs=<IPv4 CIDR>,<IPv6 CIDR>` 
      * `--feature-gates="IPv6DualStack=true"`

{{< note >}}
An IPv6 CIDR specificed via the `--cluster-cidr` flag larger than /24 will fail
{{< /note >}}

> ### Services

Reviewer 2: Users need to know that they only need to configure services cidr range for dualstackness if they plan to ingress on both `ipv4` and ipv6` services. it is optional a cluster can be dualstack egress only.

  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
```

For comparison, the following Service specification will be assigned an IPV4 address (also known as a "cluster IP") from the configured `service-cluster-ip-range` to this Service.

```yaml
apiVersion: v1
kind: Service
metadata:
  name: my-service
spec:
  ipFamily: IPv4
  selector:
    app: MyApp
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
```

#### Type LoadBalancer

On cloud providers which support IPv6 enabled external load balancers, setting the `type` field to `LoadBalancer` provisions a cloud load balancer for your Service.

### Egress Traffic

In enable IPv6 Pod routing to off-cluster destinations (eg. the Internet) the IPv6 Pod interface must be masqueraded before egressing the node. The [ip-masq-agent](https://github.com/kubernetes-incubator/ip-masq-agent) is dual stack aware and may be used to manage IP masquerading on dual stack enabled clusters.

### Provisioning a dual stack enabled cluster

The following open source tooling supports the delivery of a dual stack enabled Kubernetes cluster

   * [AKS Engine](https://github.com/Azure/aks-engine/tree/master/examples/dualstack)

> ### Ecosystem tooling

Author: Probably not. I will pull it for now

The dual stack enhancement is supported in Kubernetes core. Ecosystem tooling and projects including CoreDNS and CNI plugins may choose to implement dual stack capabilities. 

### Known Issues

   * IPv6 network block assignment is using the default IPv4 cidr block size (/24)
   * Kubenet forces IPv4,IPv6 positional reporting of IPs (--cluster-cidr)

>    * Dual stack will not fucntion if `EndpointSlice` enhancement is enabled.

Reviewer 3: nit => %s/fucntion/function

### What's Next

> Details on the implementation of this feature may be found in the [IPv4/IPv6 dual stack](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20180612-ipv4-ipv6-dual-stack.md) KEP.

Reviewer: I recommend omitting this. The What's Next section is for readers viewing whole page, not just the subset who want to learn about IPv6 / dualstack.

to run, and in both cases, the network provides one IP address per pod - as is standard for Kubernetes.

## IPv4/IPv6 dual stack

{{< feature-state state="alpha" >}}

If you enable IPv4/IPv6 dual stack networking for you Kubernetes cluster, the cluster will support the simultaneous assignment of both IPv4 and IPv6 addresses.

### Supported Features

Enabling IPv4/IPv6 dual stack on your Kubernetes cluster provides the following features:

   * Dual stack Pod networking (a single IPv4 and IPv6 address assignment per Pod)
   * IPv4 and IPv6 enabled Services (must be only of a single address family)
   * Kubenet multi address family support (IPv4 and IPv6)
   * Pod off-cluster egress routing (eg. the Internet) via both IPv4 and IPv6 interfaces

### Prerequisites

The following prerequisites are needed in order to utilize IPv4/IPv6 dual stack Kubernetes clusters:

   * Kubernetes 1.16 or later
   * Provider support for dual stack networking (Cloud provider or otherwise must be able to provide Kubernetes nodes with routable IPv4/IPv6 network interfaces)
   * Kubenet network plugin
   * Kube-proxy running in mode IPVS

### Enable IPv4/IPv6 dual stack

> To enable IPv4/IPv6 dual stack, enable the `IPv6DualStack` [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) for the relevant components of your cluster, and set dual-stack cluster network assignments:

Reviewer: nit (fixing my own typo from 7df083dc65d5c2dd1e7a52b5725483d9a306019f)
```suggestion
To enable IPv4/IPv6 dual stack, enable the `IPv6DualStack` [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) for the relevant components of your cluster, and set dual-stack cluster network assignments:
```

      * `--service-cluster-ip-range=<IPv4 CIDR>,<IPv6 CIDR>`
   * kubelet:
      * `--feature-gates="IPv6DualStack=true"`
   * kube-proxy:
      * `--proxy-mode=ipvs`
      * `--cluster-cidrs=<IPv4 CIDR>,<IPv6 CIDR>` 
      * `--feature-gates="IPv6DualStack=true"`

{{< note >}}
An IPv6 CIDR specificed via the `--cluster-cidr` flag larger than /24 will fail
{{< /note >}}

### Services

If your cluster has IPv4/IPv6 dual-stack networking enabled, you can create {{< glossary_tooltip text="Services" term_id="service" >}} with either an IPv4 or an IPv6 address. You can choose the address family for the Service's cluster IP by setting a field, `.spec.ipFamily`, on that Service.
You can only set this field when creating a new Service. The default address family for your cluster is the address family of the first service cluster IP range configured via the `--service-cluster-ip-range` flag to the kube-controller- manager.

You can set `.spec.ipFamily` to either:

   * IPv4: api-server will assign an IP from a `service-cluster-ip-range` that is `ipv4`
   * IPv6: api-server will assign an IP from a `service-cluster-ip-range` that is `ipv6`

> The following Service specification includes the `ipFamily` field. Kubernetes will assign an IPv6 address (also known as a "cluster IP") from the configured `service-cluster-ip-range` to this Service.

Reviewer 2: a default example will also be nice in the below specs.

```yaml
apiVersion: v1
kind: Service
metadata:
  name: my-service
spec:
  ipFamily: IPv6
  selector:
    app: MyApp
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
```

For comparison, the following Service specification will be assigned an IPV4 address (also known as a "cluster IP") from the configured `service-cluster-ip-range` to this Service.

```yaml
apiVersion: v1
kind: Service
metadata:
  name: my-service
spec:
  ipFamily: IPv4
  selector:
    app: MyApp
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
```

#### Type LoadBalancer

On cloud providers which support IPv6 enabled external load balancers, setting the `type` field to `LoadBalancer` provisions a cloud load balancer for your Service.

> ### Egress Traffic

Reviewer 2: That is not true. A CNI can globally reachable `IPv6` (or even `IPv4`). Masquerading is only needed if the user plan to egress on non public/routable `IPv6` addresses.

Author: Good point. I will adjust verbiage to represent that

Kubernetes does not support using globally-reachable IPv6 addresses for Pods. If you have a Pod that uses IPv6 and want that Pod to readoff-cluster destinations (eg. the public Internet), you must set up IP masquerading for the egress traffic and any replies. The [ip-masq-agent](https://github.com/kubernetes-incubator/ip-masq-agent) is dual stack aware, so you can use ip-masq-agent for IP masquerading on dual-stack clusters.

> ### Provisioning a dual stack enabled cluster

Reviewer 2: is it `dualstack` or `dual stack`?

Author: or dual-stack? Thoughts @sftim

Author: I'm going with `dual-stack`

or stand-alone.  In either version, it doesn't require any configuration or extra code
to run, and in both cases, the network provides one IP address per pod - as is standard for Kubernetes.

## IPv4/IPv6 dual-stack

{{< feature-state for_k8s_version="v1.16" state="alpha" >}}

If you enable IPv4/IPv6 dual-stack networking for your Kubernetes cluster, the cluster will support the simultaneous assignment of both IPv4 and IPv6 addresses.

### Supported Features

Enabling IPv4/IPv6 dual-stack on your Kubernetes cluster provides the following features:

   * Dual-stack Pod networking (a single IPv4 and IPv6 address assignment per Pod)

>    * IPv4 and IPv6 enabled Services (must be only of a single address family)

Reviewer: ```suggestion
   * IPv4 and IPv6 enabled Services (each Service must be for a single address family)
```

Reviewer: I think that's more readable and also easier to translate.

or stand-alone.  In either version, it doesn't require any configuration or extra code
to run, and in both cases, the network provides one IP address per pod - as is standard for Kubernetes.

## IPv4/IPv6 dual stack

{{< feature-state state="alpha" >}}

> {{< warning >}}Kubernetes' implementation of dual-stack networking is in flux. You may well find that the design is changed in future releases.{{< /warning >}}

Reviewer: The style guide [recommends](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) avoiding statements that are tied to a point in time.
The alpha feature state should serve as enough warning, I hope.

Reviewer 4: Perhaps `<feature-state state="alpha" for_k8s_version="v1.16">` would provide both a true-across-time statement, and further highlight the 1.16 requirement?

> IPv4/IPv6 dual stack enabled Kubernetes clusters allow the assignment of both IPv4 and IPv6 addresses simultaneously which enables the following capabilities:

Reviewer: ```suggestion
If you enable  IPv4/IPv6 dual stack networking for you Kubernetes cluster, the cluster will support simultaneous assignment of both IPv4 and IPv6 addresses. Building on that, you get extra capabilities:
```

   * Awareness of multiple IPv4/IPv6 address assignments per Pod

>    * Native IPv4-to-IPv4 in parallel with IPv6-to-IPv6 communications to, from, and within a cluster

Reviewer: ```suggestion
   * Native IPv4-to-IPv4 in parallel with IPv6-to-IPv6 communications to, from, and within your cluster
```

> Details on the implementation of this feature may be found in the [IPv4/IPv6 dual stack KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20180612-ipv4-ipv6-dual-stack.md)

Reviewer: I recommend moving this to a what's next section at the foot of the page. Also, move “KEP” outside of the hyperlink.

### Supported Features

> Enabling IPv4/IPv6 dual stack on a Kubernetes cluster enables the following features:

Reviewer: ```suggestion
Enabling IPv4/IPv6 dual stack on your Kubernetes cluster provides the following features:
```

Reviewer: Does this feature improve cluster operators's experience if they want to deploy Pods with multiple IPv4 addresses?
If so, there's scope for documenting that as well. I think I would log an issue to make that (potential) work visible.

Author: That enhancement is not yet available. We still only support at most a single address block from each family (IPv4 or IPv6) We can readdress this once it is

   * Dual stack Pod networking (multi-IP Pod)
   * Kubenet multi address family support (IPv4 and IPv6)

>    * Egress Pod Internet routing via IPv4/IPv6 addresses

Reviewer: ```suggestion
   * Pod egress routing to the internet via IPv4 and IPv6
```
(if that rewording is describing a different thing, it means I didn't understand the original line)

### Prerequisites

The following prerequisites are needed in order to utilize IPv4/IPv6 dual stack Kubernetes clusters:

   * Provider support for dual stack networking (Cloud provider or otherwise must be able to provide Kubernetes nodes with routable IPv4/IPv6 network interfaces)
   * Kubenet network plugin

> ### Enable IPv4/IPv6 dual stack

Reviewer: I recommend moving this (and Validate) to a new Task page.

It's OK for a Task page to be short; in fact, I think some readers will prefer that.

Reviewer: I still recommend moving this to a Task page, especially if you plan to hyperlink to it.

> To enable IPv4/IPv6 dual stack, the following flags must added to these cluster components:

Reviewer: To help the reader, link to the page on feature gates.

```suggestion
To enable IPv4/IPv6 dual stack, enable the `IPv6DualStack` [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/] for the relevant components of your cluster, and set dual-stack cluster network assignments:
```

NB: I intentionally didn't write CIDR in this suggested text; `--cluster-cidr` is a slight misnomer, as CIDR is an IPv4 concept.

```shell
kubectl get nodes k8s-linuxpool1-34450317-0 -o go-template --template='{{range .spec.podCIDRs}}{{printf "%s\n" .}}{{end}}'
10.244.1.0/24
a00:100::/24
```
There should be a single IPv4 and IPv6 CIDR allocated

Validate that the node has an IPv4 and IPv6 interface detected (replace node name with a valid node from the cluster. In this example the node name is k8s-linuxpool1-34450317-0): 
```shell
kubectl get nodes k8s-linuxpool1-34450317-0 -o go-template --template='{{range .status.addresses}}{{printf "%s: %s \n" .type .address}}{{end}}'
Hostname: k8s-linuxpool1-34450317-0
InternalIP: 10.240.0.5
InternalIP: 2001:1234:5678:9abc::5
```

Validate that a Pod has an IPv4 and IPv6 address assigned. (replace Pod name with a valid Pod on the cluster. In this example the node name is n01-79845568c8-jm9sq)
```shell
kubectl get pods n01-79845568c8-jm9sq -o go-template --template='{{range .status.podIPs}}{{printf "%s \n" .ip}}{{end}}'
10.244.1.4
a00:100::4
```

> ### Known Issues

Reviewer: The style guide [recommends](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) avoiding statements that are tied to a point in time. That's especially relevant to Concept pages; ideally, the _concept_ information can stay the same even as the details change.

It is worth mentioning the /24 default size for on IPv6 CIDR blocks; this could surprise unwary readers. I recommend using a `{{< note >}}` shortcode and placing that higher on the page.

The details about kubenet are also important. Maybe under a “Kubenet” heading?
